applications/StructuralMechanicsApplication/custom_response_functions/adjoint_conditions/adjoint_semi_analytic_base_condition.cpp

```diff
@@ -0,0 +1,397 @@
+        for (IndexType i = 0; i < r_geom.size(); ++i) {
+            const auto& r_node = r_geom[i];
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISPLACEMENT, r_node);
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ADJOINT_DISPLACEMENT, r_node);
+
+            KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_X, r_node);
+            KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Y, r_node);
+            KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Z, r_node);
+        }
+
+        return return_value;
+
+        KRATOS_CATCH( "" )
+    }
+
+    // private
+    template <class TPrimalCondition>
+    double AdjointSemiAnalyticBaseCondition<TPrimalCondition>::GetPerturbationSize(const Variable<double>& rDesignVariable, const ProcessInfo& rCurrentProcessInfo) const
+    {
+        double delta = rCurrentProcessInfo[PERTURBATION_SIZE];
+        if (rCurrentProcessInfo[ADAPT_PERTURBATION_SIZE]) {
+                delta *= this->GetPerturbationSizeModificationFactor(rDesignVariable);
+        }
+
+        KRATOS_DEBUG_ERROR_IF_NOT(delta > 0) << "The perturbation size is not > 0!";
+        return delta;
+    }
+
+    template <class TPrimalCondition>
+    double AdjointSemiAnalyticBaseCondition<TPrimalCondition>::GetPerturbationSize(const Variable<array_1d<double,3>>& rDesignVariable, const ProcessInfo& rCurrentProcessInfo) const
+    {
+        double delta = rCurrentProcessInfo[PERTURBATION_SIZE];
+        if (rCurrentProcessInfo[ADAPT_PERTURBATION_SIZE]) {
+                delta *= this->GetPerturbationSizeModificationFactor(rDesignVariable);
+        }
+
+        KRATOS_DEBUG_ERROR_IF_NOT(delta > 0) << "The perturbation size is not > 0!";
+        return delta;
+    }
+
+    template <class TPrimalCondition>
+    double AdjointSemiAnalyticBaseCondition<TPrimalCondition>::GetPerturbationSizeModificationFactor(const Variable<double>& rDesignVariable) const
+    {
+        KRATOS_TRY;
+
+        if ( mpPrimalCondition->Has(rDesignVariable) ) {
+            const double variable_value = mpPrimalCondition->GetValue(rDesignVariable);
+            return variable_value;
+        }
+        else {
+            return 1.0;
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    template <class TPrimalCondition>
+    double AdjointSemiAnalyticBaseCondition<TPrimalCondition>::GetPerturbationSizeModificationFactor(const Variable<array_1d<double,3>>& rDesignVariable) const
+    {
+        KRATOS_TRY;
+
+        // For shape derivatives the size of the geometry (length, area, ...) is used as default perturbation size modification factor.
+        // Later on this value is multiplied with a user defined factor. This product is then used as final perturbation size for computing
+        // derivatives with finite differences.
+        if(rDesignVariable == SHAPE_SENSITIVITY) {
+            const double domain_size = mpPrimalCondition->GetGeometry().DomainSize();
+            KRATOS_DEBUG_ERROR_IF(domain_size <= 0.0)
```

Comment:
I would check this only once in `Check`

Reply:
This would lead to the problem that it would be also checked for the point load which has a domain size of zero.